text/profiles/0239-profiles-data-model.md

```diff
@@ -0,0 +1,1471 @@
+  3. A request is received, the system measures 1 request.
+  4. A request is received, the system measures 1 request.
+  5. The 1 second collection cycle ends. A metric is exported for the
+      number of requests received over the interval of time t_0 to
+      t_0+1 with a value of 3.
+  6. A request is received, the system measures 1 request.
+  7. A request is received, the system measures 1 request.
+  8. The 1 second collection cycle ends. A metric is exported for the
+      number of requests received over the interval of time t_0 to
+      t_0+2 with a value of 5.
+  9. The system experiences a fault and loses state.
+  10. The system recovers and resumes receiving at time=t_1.
+  11. A request is received, the system measures 1 request.
+  12. The 1 second collection cycle ends. A metric is exported for the
+      number of requests received over the interval of time t_1 to
+      t_0+1 with a value of 1.
+
+  Note: Even though, when reporting changes since last report time, using
+  CUMULATIVE is valid, it is not recommended. */
+  AGGREGATION_TEMPORALITY_CUMULATIVE = 2;
+}
+
+// ValueType describes the semantics and measurement units of a value.
+message ValueType {
+  int64 type = 1; // Index into string table.
+  int64 unit = 2; // Index into string table.
+
+  AggregationTemporality aggregation_temporality = 3;
+}
+
+// Each Sample records values encountered in some program
+// context. The program context is typically a stack trace, perhaps
+// augmented with auxiliary information like the thread-id, some
+// indicator of a higher level request being handled etc.
+message Sample {
+  // The indices recorded here correspond to locations in Profile.location.
+  // The leaf is at location_index[0]. [deprecated]
+  repeated uint64 location_index = 1;
+  // The type and unit of each value is defined by the corresponding
+  // entry in Profile.sample_type. All samples must have the same
+  // number of values, the same as the length of Profile.sample_type.
+  // When aggregating multiple samples into a single sample, the
+  // result has a list of values that is the element-wise sum of the
+  // lists of the originals.
+  repeated int64 value = 2;
+  // label includes additional context for this sample. It can include
+  // things like a thread id, allocation size, etc.
+  //
+  // NOTE: While possible, having multiple values for the same label key is
+  // strongly discouraged and should never be used. Most tools (e.g. pprof) do
+  // not have good (or any) support for multi-value labels. And an even more
+  // discouraged case is having a string label and a numeric label of the same
+  // name on a sample.  Again, possible to express, but should not be used.
+  // [deprecated, superseded by attributes]
+  repeated Label label = 3;
+
+  // locations_start_index along with locations_length refers to to a slice of locations in Profile.location. Supersedes location_index.
+  uint64 locations_start_index = 4;
```

Comment:
I'm still not sure this single table of locations and pointing to a range in it from samples is worth it. I wonder if a separate Callstacks table that would contain specific stacks and then referring to it by index would be simpler to understand and not much less efficient.

The downsides of the current approach as I see them are:
* We are slicing a single array so it's effectively a little of custom memory management which means things like array-out-of-bounds can go unnoticed.
* Writing code for profile.Merge() is going to be tricky as it's not obvious what it should optimize for.
* Collector code usually needs to be as simple as possible, so putting there any heuristics for trying to minimize the locations table is probably unlikely.

Reply:
Also, if we do that, then we'd have a much more natural place to put `stacktrace_id_index` into. I think there it could even be just a string field, wouldn't have to be an index since it would be unique per unique stack trace anyway.

Reply:
I mean something like [this](https://github.com/google/pprof/compare/main...aalexand:pprof:otel-proto?expand=1).

Reply:
We tried that in the past (For reference, we discussed it on October 5, I explained it around 23:00 mark) and unfortunately, that increases the number of objects allocated quite a bit, and is less efficient compared to the version where locations are in a shared array.

Here are the pseudocode diff between 2 versions:
* Oct 4 (pseudocode):
```proto
struct Profile {
  repeated Stacktrace stacktraces = 1;
  repeated Sample samples = 2;
}

struct Sample {
  uint64 stacktrace_id = 1;
}

struct Stacktrace {
  uint64 location_id = 1;
}
```


* After Stacktrace removal (Oct 6 2023) (pseudocode)
```proto
struct Profile {
  repeated uint64 location_ids = 1;
  repeated Sample samples = 2;
}

struct Sample {
  uint64 location_start_index = 1;
  uint64 location_length = 2;
}
```

And here's a screenshot from benchmarks:
![Screenshot 2024-01-03 at 8 27 43 PM](https://github.com/open-telemetry/oteps/assets/662636/6045a33a-25de-4041-b712-65cf796f4e7b)

[Link to the benchmarks](https://docs.google.com/spreadsheets/d/1Q-6MlegV8xLYdz5WD5iPxQU2tsfodX1-CDV1WeGzyQ0/edit?usp=sharing) ("Oct 4 vs Oct 6 After Stacktrace removal" sheet)